packages/amplify_core/lib/src/types/models/mipr/model_association.dart

```diff
@@ -0,0 +1,145 @@
+import 'package:aws_common/aws_common.dart';
+import 'package:built_collection/built_collection.dart';
+import 'package:built_value/built_value.dart';
+import 'package:built_value/serializer.dart';
+
+part 'model_association.g.dart';
+
+/// {@template amplify_core.models.mipr.model_association}
+/// The type of association/relationship of a model's field as defined by one
+/// of the Amplify relationship directives:
+///
+/// - `@hasOne`: [hasOne]
+/// - `@hasMany`: [hasMany]
+/// - `@belongsTo`: [belongsTo]
+/// - `@manyToMany`: [manyToMany]
+/// {@endtemplate}
+class ModelAssociationType extends EnumClass {
+  const ModelAssociationType._(super.name);
+
+  /// An `@hasOne` association.
+  ///
+  /// Indicates a model is in a 1->1 relationship with the associated model.
+  /// In this relationship, the foreign key is owned by the associated model.
+  @BuiltValueEnumConst(wireName: 'HasOne')
+  static const ModelAssociationType hasOne = _$hasOne;
+
+  /// An `@hasMany` association.
+  ///
+  /// Indicates a model is in a 1->N relationship with the associated model.
+  /// In this relationship, the foreign key is owned by the associated model.
+  @BuiltValueEnumConst(wireName: 'HasMany')
+  static const ModelAssociationType hasMany = _$hasMany;
+
+  /// An `@belongsTo` association.
+  ///
+  /// Indicates a model is in a 1->1 relationship with the associated model.
+  /// In this relationship, the foreign key is owned by this model.
+  @BuiltValueEnumConst(wireName: 'BelongsTo')
+  static const ModelAssociationType belongsTo = _$belongsTo;
+
+  /// An `@manyToMany` association.
+  ///
+  /// Indicates a model is in a N->N relationship with the associated model.
+  /// In this relationship, the foreign keys are stored in a join table.
+  @BuiltValueEnumConst(wireName: 'ManyToMany')
+  static const ModelAssociationType manyToMany = _$manyToMany;
+
+  /// The values of [ModelAssociationType].
+  static BuiltSet<ModelAssociationType> get values =>
+      _$ModelAssociationTypeValues;
+
+  /// The [ModelAssociationType] value corresponding to [name].
+  static ModelAssociationType valueOf(String name) =>
+      _$ModelAssociationTypeValueOf(name);
+
+  /// The serializer for [ModelAssociationType].
+  static Serializer<ModelAssociationType> get serializer =>
+      _$modelAssociationTypeSerializer;
+}
+
+/// {@template amplify_core.models.mipr.model_association}
+/// Represents an association/relationship between two model types.
+/// {@endtemplate}
+abstract class ModelAssociation
+    with AWSSerializable<Map<String, Object?>>
+    implements Built<ModelAssociation, ModelAssociationBuilder> {
+  /// {@macro amplify_core.models.mipr.model_association}
+  factory ModelAssociation({
+    required ModelAssociationType associationType,
+    String? targetName,
+    String? associatedName,
+    String? associatedType,
+  }) =>
+      _$ModelAssociation._(
+        associationType: associationType,
+        targetName: targetName,
```

Comment:
`targetName` should be deprecated from codegen in favor of CPK implementation. I don't think we need to keep this field for the new gen of DataStore?

Reply:
Okay, I was wondering about that. Sounds good

Reply:
Note: Some tests will fail now, but I think it's probably better to decide on the interface we want nwo.